### PR TITLE
Add attention mechanism into seq2seq example

### PR DESCRIPTION
**Pull request by soskek**
Add a feature of attention mechanism into the current seq2seq example.
https://github.com/chainer/chainer/issues/3799

Issue is from https://github.com/chainer/chainer/issues/4122

### CODE REVIEW

**Comment by delta2323** [comment]
@soskek Can I help you for this PR?

**Comment by soskek** [comment]
Thank you!
I will send updated PR in late Jan. Then, could you review it, though @unnonouno will be also good?

**Comment by delta2323** [comment]
> Then, could you review it, though @unnonouno will be also good?

Either is OK. I can take over this PR to @unnonouno. It is up to him.

**Comment by soskek** [comment]
jenkins test this please

**Comment by soskek** [comment]
I've done it.
- Implement the multiplicative attention mechanism (as option)
- Implement the bi-LSTM encoder (as option)
- Change hyperparameters to better ones by reference to [a paper](https://arxiv.org/pdf/1703.03906.pdf)
- Minor fix of printing results

**Comment by delta2323** [comment]
Good! Could you remove [WIP] if the initial work is done?

**Comment by delta2323** [comment]
@unnonouno Should I take over the PR to you?

File: examples/seq2seq/seq2seq.py [hunk 1]
```diff
@@ -254,14 +389,24 @@ def main():
     target_words = {i: w for w, i in target_ids.items()}
     source_words = {i: w for w, i in source_ids.items()}
 
-    # Setup model
+
+<< << << < HEAD
```
Comment (unnonouno) [inline examples/seq2seq/seq2seq.py]: woops

File: examples/seq2seq/seq2seq.py [hunk 2]
```diff
@@ -27,19 +27,125 @@ def sequence_embed(embed, xs):
...
+        concat_K, k_pad_mask = prepare_attention(ks)
+        batchsize, k_len, k_units = concat_K.shape
+        K = self.W_key(concat_K.reshape(batch * k_len, k_units))
+        K = K.reshape(batch, k_len, self.att_units)
+        assert K.shape == (batch, k_len, self.att_units)
+
+        QK_dot = F.batch_matmul(Q, K, transb=True)
+        assert QK_dot.shape == (batch, q_len, k_len)
+        # ignore attention weights where padded values are used
+        QK_pad_mask = q_pad_mask[:, :, None] * k_pad_mask[:, None, :]
```
[the remark this answers is not in the document]
Comment (soskek) [reply examples/seq2seq/seq2seq.py]: This multiplication is performed with padding masks whose values are 1 or 0. So, this is OK.

File: examples/seq2seq/seq2seq.py [hunk 3]
```diff
@@ -27,19 +27,125 @@ def sequence_embed(embed, xs):
...
+class AttentionMechanism(chainer.Chain):
+    def __init__(self, n_units):
+        super(AttentionMechanism, self).__init__()
+        self.n_units = n_units
+        self.att_units = min(256, n_units)
+        with self.init_scope():
+            self.W_query = L.Linear(None, self.att_units)
+            self.W_key = L.Linear(None, self.att_units)
+
+    def __call__(self, qs, ks):
```
[the remark this answers is not in the document]
Comment (soskek) [reply examples/seq2seq/seq2seq.py]: OK.

File: examples/seq2seq/seq2seq.py [hunk 4]
```diff
@@ -27,19 +27,125 @@ def sequence_embed(embed, xs):
...
+    split_V = F.split_axis(V.reshape(batch * max_len, units),
+                           stitch_split_ids, axis=0)[::2]
+    return split_V
+
+
+class AttentionMechanism(chainer.Chain):
+    def __init__(self, n_units):
+        super(AttentionMechanism, self).__init__()
+        self.n_units = n_units
+        self.att_units = min(256, n_units)
```
[the remark this answers is not in the document]
Comment (soskek) [reply examples/seq2seq/seq2seq.py]: This is typically the same as `n_units`. I added the argument which copies `n_units` by default.

**Comment by soskek** [comment]
@unnonouno Please re-review this.

**Comment by crcrpar** [comment]
bump

**Pull request closed without merging**